src/libOpenImageIO/imageoutput.cpp

```diff
@@ -261,17 +262,13 @@ ImageOutput::append_error(string_view message) const
 {
     if (message.size() && message.back() == '\n')
         message.remove_suffix(1);
-    std::string* errptr = m_impl->m_errormessage.get();
-    if (!errptr) {
-        errptr = new std::string;
-        m_impl->m_errormessage.reset(errptr);
-    }
+    std::string& err_str = error_messages[this];
```

Comment:
So there is a per-thread map of ImageOutput -> string. This line adds a slot in this thread's map for this ImageOutput. When the thread terminates, I assume the whole map (for that thread) will delete. But as long as the thread is alive, the strings just sit there even when the ImageOutput they correspond to is long gone, no? Don't we want something in `ImageOutput::Impl::~Impl` that does an `erase` for its slot in the map?

Reply:
The problem is the destructor only runs on one thread. So it can't reach into other thread's storage without additional bookkeeping.

It probably is solveable, though I worry about introducing complicated threaded bookkeeping just to release the memory a few error messages.

Reply:
Oooooh, yes of course.

OK, so for the error things, it might pseudo-leak a few strings -- and only if there are actual errors. So in practice, NBD.

And for ImageCache, the "object" is the cache itself, of which there is generally exactly one, so we don't really worry about a *succession* of IC's being created from scratch, then being destroyed without freeing their per-thread cruft.

Sounds ok to me, then.

Reply:
Chris, I still think that maybe the object destructors should have a

`if (threadlocal.contains(key)) threadlocal.erase(key);`

It's true that it will only free the thing if the same thread that's destroying the ImageOutput (say) is the same one that called error(). But that's a very common case!  So we could still leak a few strings, but usually it will cause us to fully clean up.

Reply:
I can definitely add this to all the destructors. It shouldn't cost much in the common case and keeps the map for the main thread small(er).

I also just realized that technically, these error strings have the same bug I fixed in the image cache. It's much less likely to cause issues, but the bug would be:

1) create an `ImageInput` - gets allocated at address 0x1234 (for example)
2) do some operations (potentially across threads) which lead to errors, but you don't check or clear them
3) close the `ImageInput`
4) create a new `ImageInput` which happens to get allocated at 0x1234 again
5) now all threads still have the error message from the first one which never got cleared :(


The way I solved it for the per thread infos was to make the map use a strictly incrementing counter instead of their address for the lookup (in practice most apps only ever create a single image cache so its not a big diference).  I think I should do the same for the error messages (even though its asymptomatic in the testsuite, the potentialy for confusion if it ever happens would be quite high). The only drawback is an extra 64 bit ID per `ImageInput`,`ImageOutput`.

I'll ponder this a bit more to see if I think of any other clever way to solve this ....

Reply:
haha, I think our messages crossed, I had typed mine a couple hours ago, but only hit the button now without seeing you commented on the same thing.

Reply:
In short, I don't think it's worth doing anything elaborate. Let's consider it a user error to destroy an ImageInput that you used, accumulated errors, but never retrieved the error messages. But if you make the destructor clear the error for that thread (which will usually be the only thread), I think the stars will rarely align to give a problem, even when the user makes that mistake.